File: Binary Tree/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = float("-inf")
        
        def findMaxPath(root):
            nonlocal res
            if root is None:
                return 0 
            left = findMaxPath(root.left)
            right = findMaxPath(root.right)
            # evaluate max sum here 
            
            if left <= 0 and right <= 0:
                res = max(res, root.val)
                
                return root.val
            elif left <= 0 and right >= 0:
                res = max(root.val + right, res)
                
                return root.val + right 
            elif left >= 0 and right <= 0:
                res = max(root.val + left, res)
                return root.val + left 
            elif left >= 0 and right >= 0:
                res = max(left + right + root.val , res)
                return max(left, right) + root.val
            
        findMaxPath(root)
        return res

File: Binary Tree/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int res = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        // every node will return
        dfs(root);
        return res;
    }
    int dfs(TreeNode* root) {
        if (!root) return -2000;

        int left = dfs(root->left);
        int right = dfs(root->right);
        vector<int> comp = {res, root->val, left + root->val + right, root->val + left, root->val + right};
        vector<int> ret = {root->val, root->val + left, root->val + right};
        res = *max_element(comp.begin(), comp.end());

        return *max_element(ret.begin(), ret.end());
    }
};